### PR TITLE
Added an actual AUR cache

### PR DESCRIPTION
**Pull request by jdonszelmann**
Also added a simple caption on the main table if it's empty

### CODE REVIEW

File: pkg/store/badger_cache.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/finitum/AAAAA/pkg/aur"
+	"github.com/pkg/errors"
+	"time"
+)
+
+type BadgerCache struct {
+	db *badger.DB
+}
+
+const userPrefix = "user_"
```
Comment (NULLx76) [inline pkg/store/badger_cache.go]: Why is user prefix defined in the cache file and vice versa?,

File: pkg/store/badger_cache.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+const cacheTTL = 30 * time.Minute
+
+func OpenBadgerCache(path string) (*BadgerCache, error) {
+	db, err := badger.Open(badger.DefaultOptions(path + ".cache"))
+	if err != nil {
+		return nil, errors.Wrap(err, "opening badger store")
+	}
+
+	// TODO: Schedule garbage collection
```
Comment (NULLx76) [inline pkg/store/badger_cache.go]: I was under the impression that badger did this itself, is that incorrect?

File: pkg/store/badger_cache.go (hunk 1 above)
Comment (jdonszelmann) [reply pkg/store/badger_cache.go]: copied the wrong thing, oops

File: pkg/store/badger_cache.go (hunk 2 above)
Comment (jdonszelmann) [reply pkg/store/badger_cache.go]: that's incorrect. You have to call the GC function or close the DB (that also collects garbage)

Comment (jdonszelmann) [reply pkg/store/badger_cache.go]: https://dgraph.io/docs/badger/get-started/#garbage-collection

File: pkg/store/cache.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package store
+
+import "github.com/finitum/AAAAA/pkg/aur"
+
+type AurCache interface {
+	SetEntry(searchterm string, result aur.Results) error
+	GetEntry(searchterm string) (aur.Results, bool, error)
+}
+
+type Cache interface {
```
Comment (NULLx76) [inline pkg/store/cache.go]: This type seems useless?

Comment (jdonszelmann) [reply pkg/store/cache.go]: removed it

File: services/aur_cache/main.go [hunk 4]
```diff
@@ -4,39 +4,85 @@ import (
...
-
 	r.Use(render.SetContentType(render.ContentTypeJSON))
 
-	r.Get("/search/{term}", proxy)
+	cache, err := store.OpenBadgerCache(os.TempDir() + "/AAAAA")
+	if err != nil {
+		log.Fatalf("Couldn't open store (%v)", err)
+	}
+
+	r.Get("/search/{term}", func(writer http.ResponseWriter, request *http.Request) {
```
Comment (NULLx76) [inline services/aur_cache/main.go]: If you make proxy return a handler this can be made cleaner

File: services/aur_cache/main.go [hunk 5]
```diff
@@ -4,39 +4,85 @@ import (
...
+func proxy(cache store.Cache, w http.ResponseWriter, r *http.Request) {
 	term := chi.URLParam(r, "term")
 	if len(term) < 3 {
 		http.Error(w, "term too short", http.StatusBadRequest)
 		return
 	}
 
+	cachedResult, exact, err := cache.GetEntry(term)
+	if err == nil {
+		log.Info("Cache hit!")
```
Comment (NULLx76) [inline services/aur_cache/main.go]: Trace

File: services/aur_cache/main.go (hunk 4 above)
Comment (jdonszelmann) [reply services/aur_cache/main.go]: what do you mean?

File: services/aur_cache/main.go (hunk 5 above)
Comment (jdonszelmann) [reply services/aur_cache/main.go]: fixed

File: services/aur_cache/main.go [hunk 6]
```diff
@@ -4,39 +4,85 @@ import (
...
+
+		_ = render.Render(w, r, filteredResult)
+		return
+	}
+
+	// otherwise it's just a cache miss
+	if err != store.ErrNotExists {
+		log.Errorf("An unexpected error occurred while retrieving cache results. Attempting a non-cached lookup (%v)", err)
+	} else {
+		log.Info("Cache miss!")
```
Comment (NULLx76) [inline services/aur_cache/main.go]: Trace

Comment (jdonszelmann) [reply services/aur_cache/main.go]: yup, fixed together with the one above

File: pkg/store/badger_cache.go (hunk 2 above)
Comment (NULLx76) [reply pkg/store/badger_cache.go]: We should do that then

**Pull request merged**